politeiad/backend/gitbe/cms.go

```diff
@@ -966,6 +966,196 @@ func (g *gitBackEnd) tallyDCCVotes(token string) ([]cmsplugin.CastVote, error) {
 	return cv, nil
 }
 
+// pluginDCCVoteDetails returns the VoteDetails of a requested DCC vote.
+// It uses the caches that should be populated with the StartVotes and
+// StartVoteReplies.
+func (g *gitBackEnd) pluginDCCVoteDetails(payload string) (string, error) {
+	log.Tracef("pluginDCCVoteDetails: %v", payload)
+
+	vd, err := cmsplugin.DecodeVoteDetails([]byte(payload))
+	if err != nil {
+		return "", fmt.Errorf("DecodeVoteResults %v", err)
+	}
+
+	g.Lock()
+	defer g.Unlock()
+
+	// git checkout master
+	err = g.gitCheckout(g.vetted, "master")
+	if err != nil {
+		return "", err
+	}
+
+	var svr cmsplugin.StartVoteReply
+	var sv cmsplugin.StartVote
+	var vdr cmsplugin.VoteDetailsReply
+	// Prepare reply
+	var (
+		dd *json.Decoder
+		ff *os.File
+	)
+	// Fill out vote
+	filename := mdFilename(g.vetted, vd.Token,
+		cmsplugin.MDStreamVoteBits)
+	ff, err = os.Open(filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			goto nodata
+		}
+		return "", err
+	}
+	defer ff.Close()
+	dd = json.NewDecoder(ff)
+
+	err = dd.Decode(&sv)
+	if err != nil {
+		if err == io.EOF {
+			goto nodata
+		}
+		return "", err
+	}
+
+	// Load the vote snapshot from disk
+	ff, err = os.Open(mdFilename(g.vetted, vd.Token,
+		cmsplugin.MDStreamVoteSnapshot))
+	if err != nil {
+		return "", err
+	}
+	defer ff.Close()
+
+	dd = json.NewDecoder(ff)
+	err = dd.Decode(&svr)
+	if err != nil {
+		return "", err
+	}
+	vdr = cmsplugin.VoteDetailsReply{
+		StartVote:      sv,
+		StartVoteReply: svr,
+	}
+
+nodata:
+	reply, err := cmsplugin.EncodeVoteDetailsReply(vdr)
+	if err != nil {
+		return "", fmt.Errorf("Could not encode VoteResultsReply: %v",
+			err)
+	}
+	return string(reply), nil
+}
+
+// pluginDCCVoteSummary
+func (g *gitBackEnd) pluginDCCVoteSummary(payload string) (string, error) {
+	log.Tracef("pluginDCCVoteSummary: %v", payload)
+
+	vs, err := cmsplugin.DecodeVoteSummary([]byte(payload))
+	if err != nil {
+		return "", fmt.Errorf("DecodeVoteResults %v", err)
+	}
+
+	// Verify proposal exists, we can run this lockless
+	if !g.vettedPropExists(vs.Token) {
+		return "", fmt.Errorf("dcc not found: %v", vs.Token)
+	}
+
+	// This portion is must run locked
+
+	g.Lock()
+	defer g.Unlock()
+
+	// Prepare reply
+	var vrr cmsplugin.VoteResultsReply
+	var vsr cmsplugin.VoteSummaryReply
+	var svr cmsplugin.StartVoteReply
+	vors := make([]cmsplugin.VoteOptionResult, 0,
+		len(vrr.StartVote.Vote.Options))
+
+	// Fill out cast votes
+	vrr.CastVotes, err = g.tallyDCCVotes(vs.Token)
+	if err != nil {
+		return "", fmt.Errorf("Could not tally votes: %v", err)
+	}
+
+	// git checkout master
+	err = g.gitCheckout(g.vetted, "master")
```

Comment:
Plugins should use the `Backend` api when possible instead of interacting with git and the filesystem directly.  For metadata streams like in this function you should be able retrieve the full record with `GetVetted` and then pull out the mdstreams that you need from `Record.Metadata`.  The same applies to the `DCCVoteDetails` function.

The only time a plugin should interact with the filesystem directly is when it needs journal data, since there aren't any `Backend` methods for interacting with the journals.